src/map/utils/battleutils.cpp

```diff
@@ -1904,6 +1904,129 @@ uint32 TakePhysicalDamage(CBattleEntity* PAttacker, CBattleEntity* PDefender, in
+                ((CPetEntity*)PDefender)->loc.zone->PushPacket(PDefender, CHAR_INRANGE, new CEntityUpdatePacket(PDefender, ENTITY_UPDATE, UPDATE_COMBAT));
+
+            break;
+
+        case TYPE_PET:
+            ((CPetEntity*)PDefender)->loc.zone->PushPacket(PDefender, CHAR_INRANGE, new CEntityUpdatePacket(PDefender, ENTITY_UPDATE, UPDATE_COMBAT));
+            break;
+        }
+
+        // try to interrupt spell
+        if (PDefender->PBattleAI->m_PMagicState)
+            PDefender->PBattleAI->m_PMagicState->TryHitInterrupt(PAttacker);
+        else
+            ShowError("battleutils::TakeWeaponskillDamage Entity (%d) has no magic state\n", PDefender->id);
+
+        int16 baseTp = 0;
+
+        if (isRanged && PAttacker->objtype == TYPE_PC)
+        {
+            int16 delay = PAttacker->GetRangedWeaponDelay(true);
+            baseTp = CalculateBaseTP((delay * 110) / 1000);
+        }
+        else
+        {
+            int16 delay = PAttacker->GetWeaponDelay(true);
+
+            if (PAttacker->m_Weapons[SLOT_SUB]->getDmgType() > 0 &&
+                PAttacker->m_Weapons[SLOT_SUB]->getDmgType() < 4 &&
+                PAttacker->m_Weapons[slot]->getDmgType() != DAMAGE_HTH)
+            {
+                delay /= 2;
+            }
+
+            float ratio = 1.0f;
+
+            if (PAttacker->m_Weapons[slot]->getDmgType() == DAMAGE_HTH)
+                ratio = 2.0f;
+
+            baseTp = CalculateBaseTP((delay * 60) / 1000) / ratio;
+        }
+
+
+        // add tp to attacker
+        PAttacker->addTP(tpMultiplier * (baseTp * (1.0f + 0.01f * (float)((PAttacker->getMod(MOD_STORETP) + getStoreTPbonusFromMerit(PAttacker))))));
+
+        //account for attacker's subtle blow which reduces the baseTP gain for the defender
+        float sBlowMult = ((100.0f - dsp_cap((float)PAttacker->getMod(MOD_SUBTLE_BLOW), 0.0f, 50.0f)) / 100.0f);
+
+        //mobs hit get basetp+30 whereas pcs hit get basetp/3
+        if (PDefender->objtype == TYPE_PC)
+            PDefender->addTP((baseTp / 3) * sBlowMult * (1.0f + 0.01f * (float)((PDefender->getMod(MOD_STORETP) + getStoreTPbonusFromMerit(PAttacker))))); //yup store tp counts on hits taken too!
+        else
+            PDefender->addTP((baseTp + 30) * sBlowMult * (1.0f + 0.01f * (float)PDefender->getMod(MOD_STORETP))); //subtle blow also reduces the "+30" on mob tp gain
+
+        if (PAttacker->objtype == TYPE_PC)
+            charutils::UpdateHealth((CCharEntity*)PAttacker);
+    }
+    else if (PDefender->objtype == TYPE_MOB)
+        ((CMobEntity*)PDefender)->PEnmityContainer->UpdateEnmityFromDamage(PAttacker, 0);
+
+
+    if (PDefender->objtype == TYPE_PC)
+        charutils::UpdateHealth((CCharEntity*)PDefender);
+
+    if (PAttacker->objtype == TYPE_PC && !isRanged)
+        PAttacker->StatusEffectContainer->DelStatusEffectsByFlag(EFFECTFLAG_ATTACK);
+
+    if (damage < 0)
+        return -damage;
+    else
+        return damage;

```

Comment:
@teschnei looking at this today, I'm thinking maybe I should change the new function to returned signed damage in case the mob absorbs the magical weaponskill. thoughts?